### ISSUE TITLE
Example for type-generic extension?

### ISSUE THREAD

**Issue by lucasb-eyer** [pytorch/extension-ffi#8]
I'm currently writing a C extension and the existing example is of good help, thanks for that already!

I was wondering if it's possible to write type-generic extensions similar to the code in `nn`, using constructs like `THTensor_(size)(x, 1)` and friends.

I think it's not obvious, because the extension-code's header file can't contain preprocessor macros. So my current guess is that one would need to manually list all signatures for all types in the header, but could at least run code-generating macros in the implementation `.c` file. Is that correct? Is including such an example planned sometime soon?

**Comment by soumith** [comment]
see pytorch/audio

**Comment by zhanghang1989** [comment]
What's the easiest way to use those type-generic extensions with `CUDA kernel functions` in CFFI? 
Thanks!

**Comment by soumith** [comment]
we need to fix header generation, but the.h needs to be pregen, as i write the command in the headetfilr th_sox.h

**Comment by soumith** [comment]
@zhanghang for .cu files you need a cmakelists file as well

**Comment by zhanghang1989** [comment]
Thanks @soumith ! I will try to figure it out.

**Comment by lucasb-eyer** [comment]
Awesome, thanks for the pointer! Looks pretty much how I imagined it would :smile:

**Comment by zhanghang1989** [comment]
Hi @soumith 
I am writing a package with generic files. However, cmake `FIND_PACKAGE(Torch REQUIRED)` finds Torch instead of PyTorch. Any comments?

**Comment by apaszke** [comment]
If you're building an extension using our ffi interface then the appropriate paths will be automatically added as include directories (you can just add `#include <TH/TH.h>`). If you're building another lib, a hacky way to find the headers would be to do this:
```python
import os
import torch
package_base = os.path.dirname(torch.__file__)
include_path = os.path.join(package_base, 'lib', 'include')
```
We should expose it as part of ffi utils...

**Comment by zhanghang1989** [comment]
Thanks @apaszke ! As far as I understand, the ffi interface doesn't support kernel function currently.

**Comment by apaszke** [comment]
Yes that's right. That's why I showed you the workaround

**Comment by zhanghang1989** [comment]
Thanks @apaszke ! I manually add the include dirs and link the libraries. It works now.

**Comment by zhanghang1989** [comment]
Hi @apaszke , question again :) How to link a library to the ffi package? Is this the right way?
https://github.com/zhanghang1989/PyTorch-Encoding-Layer/blob/master/build.py#L41
I got error `Library not loaded: @rpath/libENCODING.dylib`

Thanks!

**Comment by apaszke** [comment]
No, you should use the `libraries=` argument (see e.g. [setup.py in PyTorch repo](https://github.com/pytorch/pytorch/blob/master/setup.py#L378)).

**Comment by zhanghang1989** [comment]
Thanks a lot! It is working nicely [build.py](https://github.com/zhanghang1989/PyTorch-Encoding-Layer/blob/master/build.py#L21)

**Comment by phalexo** [comment]
I have been struggling with building some extensions within the mask-rcnn.pytorch repo. The problem appears to stem from inability to locate <TH/TH.h> and <THC/THC.,h> 

Any suggestions would be appreciated.  

pytorch itself was built/installed from its source repo under anaconda with python 3.5.
